python/ray/tune/trial_executor.py

```diff
@@ -15,32 +16,53 @@
 logger = logging.getLogger(__name__)
 
 
+# Ideally we want to use @cache; but it's only available for python 3.9.
+# Caching is only helpful for no autoscaler case.
 @lru_cache()
-def _get_warning_threshold() -> float:
+def _get_cluster_resources_no_autoscaler() -> Dict:
+    return ray.cluster_resources()
```

Comment:
When does this get updated? It would be good to refresh this with some frequency

Reply:
Hmm, it won't get refreshed in this case unless the key is evicted from LRU cache.
What is the concern here? This is only used for no autoscaler case. So resources *should* not be changing?
Some refresh policy with TTL adds complexity but also can be hard to work with - one is stuck with staled data anyways.

Reply:
Even without autoscaler, nodes may take a bit of time before they are fully added to the cluster (or nodes may be lost). In that situation, it's ok to have some staleness, but it's important to bound it. 

For example, if a worker node dies, your tune job should be aware (though the SLA is not too strict).

Reply:
Hmm, I see. Cluster resources is dynamic even without autoscaler. 
So in this particular case, should we then delay the checking logic?
I wouldn't worry too much about nodes getting lost (in which case cluster_resources returns more than what's available); but it would be pretty bad if cluster_resources returns less than what is possible due to nodes still getting added and we prematurely throw a user visible TuneError, with instructions that clear don't match what happened and user would have no way of bypassing this issue. Thoughts?

Reply:
Talked offline. Elongated the threshold to check for resources to be 10s (no autoscaler), under the assumption that by 10s, ray.cluster_resources() should give us a stable view of cluster resources (it can become less if nodes are not available but that's not relevant to the logic here).